### DISCUSSION TITLE
Problem with debug data

### DISCUSSION THREAD

**Discussion by tohidmonjazeb** [explosion/spaCy#9759] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

python -m spacy debug data ./config.cfg
## Your Environment

* Operating System: windows 10
* Python Version Used: 3.8.5
* spaCy Version Used: 3.1.3
* Environment Information:
* I have made train.spacy and dev.spacy files then I ran "python -m spacy debug data ./config.cfg" and nothing happened
I mean that it just shows :

2021-11-27 13:49:32.956042: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2021-11-27 13:49:32.956288: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.

============================ Data file validation ============================

And nothing else even when I wait for 1 hour.
**Could you please help me with this?**

**Comment by polm** [comment, marked as answer]
`spacy debug data` loads the data into memory by default. After one hour I'd expect you to have run out of memory, but maybe something weird is going on. How big is the data you're passing to `debug data`, in file size, number of docs, number of words? Can you share your config and the output of `spacy info`?

**Reply by tohidmonjazeb** [reply]
@polm 
train.spacy and dev.spacy files are 5 and 1 Megabytes , 26000 and 1500 docs , 460000 and 25000 words.

spacy info output : 

============================== Info about spaCy ==============================

spaCy version    3.1.3
Location         C:\Users\SETUP CO\anaconda3.2\lib\site-packages\spacy
Platform         Windows-10-10.0.19041-SP0
Python version   3.8.5
Pipelines        en_core_web_sm (3.1.0)

my config : [https://github.com/tohidmonjazeb/spacy_config/blob/main/config.cfg](url)

**Reply by polm** [reply]
Hm, 460k words is a lot, but not *that* much. It's hard to say what's going on with just the information here. 

Could you try running with a subset of your data, say like 10%, and see if that works? That way we would at least know it was size related.

As another option, would it be possible for you to provide us some sample data to check? It's fine if you anonymize it by replacing all the words with "xxx" or whatever.

**Reply by tohidmonjazeb** [reply]
I just have tried what you said and got the same result (with 10%), nothing happens after waiting for 20 minutes.
for that another option, Could you please tell me what file you want me to share with you ? Do you mean .spacy files ?

**Reply by polm** [reply]
Yes, if you can give us spaCy files we can try to reproduce the issue.

**Reply by tohidmonjazeb** [reply]
data and config : https://github.com/tohidmonjazeb/spacy_config

**Reply by polm** [reply]
Thanks! It looks like dep annotations have a problem and your data is full of cycles. `debug data` shouldn't hang in this situation, but I don't think we've tested it much with invalid data, so that's a bug we need to fix, but your data isn't usable in its current state.

I'm not sure how you generated your annotations, but you seem to have many situations where two words, A and B, will be each other's heads, which is incorrect. The dependency relations should form an upside-down tree with one root.

**Reply by polm** [reply]
Actually, one small note - while you do have the situation I described where two nodes form a cycle often, it looks like even more often you have a situation where a word is its own head. That should only be the case for the root in the sentence (so just one word).

**Reply by tohidmonjazeb** [reply]
Actually I got this data set from https://github.com/UniversalDependencies/UD_Persian-PerDT 
after examining it  I faced with many wrong annotations, so I tried to fix them but it seems that I should not use this data at all.
Anyway, I'm so thankful for your guidance and responsibility .

**Reply by adrianeboyd** [reply]
We provide converters for the UD CoNLL-U format. Since this UD corpus contains multi-word tokens, which spacy `Doc`s don't support, I'd usually recommend converting it like this (every ten sentences in a paragraph, merge multi-word tokens to tokens that spacy supports):

```bash
python -m spacy convert -n 10 -T fa_perdt-ud-train.conllu .
```

**Reply by tohidmonjazeb** [reply]
Actually I didn't use spacy's converter because this data contains many null values and also some wrong annotations with heads.
so I preprocessed it manually and made Doc objects and added them to DocBin and then made spacy files from them.
